node-js-basics-101/node_modules/ws/lib/PerMessageDeflate.js

```diff
@@ -0,0 +1,384 @@
+   * @public
+   */
+  decompress (data, fin, callback) {
+    const endpoint = this._isServer ? 'client' : 'server';
+
+    if (!this._inflate) {
+      const maxWindowBits = this.params[`${endpoint}_max_window_bits`];
+      this._inflate = zlib.createInflateRaw({
+        windowBits: typeof maxWindowBits === 'number' ? maxWindowBits : DEFAULT_WINDOW_BITS
+      });
+    }
+    this._inflate.writeInProgress = true;
+
+    var totalLength = 0;
+    const buffers = [];
+    var err;
+
+    const onData = (data) => {
+      totalLength += data.length;
+      if (this._maxPayload < 1 || totalLength <= this._maxPayload) {
+        return buffers.push(data);
+      }
+
+      err = new Error('max payload size exceeded');
+      err.closeCode = 1009;
+      this._inflate.reset();
+    };
+
+    const onError = (err) => {
+      cleanup();
+      callback(err);
+    };
+
+    const cleanup = () => {
+      if (!this._inflate) return;
+
+      this._inflate.removeListener('error', onError);
+      this._inflate.removeListener('data', onData);
+      this._inflate.writeInProgress = false;
+
+      if (
+        (fin && this.params[`${endpoint}_no_context_takeover`]) ||
+        this._inflate.pendingClose
+      ) {
+        this._inflate.close();
+        this._inflate = null;
+      }
+    };
+
+    this._inflate.on('error', onError).on('data', onData);
+    this._inflate.write(data);
+    if (fin) this._inflate.write(TRAILER);
+
+    this._inflate.flush(() => {
+      cleanup();
+      if (err) callback(err);
+      else callback(null, bufferUtil.concat(buffers, totalLength));
+    });
+  }
+
+  /**
+   * Compress data.
+   *
+   * @param {Buffer} data Data to compress
+   * @param {Boolean} fin Specifies whether or not this is the last fragment
+   * @param {Function} callback Callback
+   * @public
+   */
+  compress (data, fin, callback) {
+    if (!data || data.length === 0) {
+      process.nextTick(callback, null, EMPTY_BLOCK);
+      return;
+    }
+
+    const endpoint = this._isServer ? 'server' : 'client';
+
+    if (!this._deflate) {
+      const maxWindowBits = this.params[`${endpoint}_max_window_bits`];
+      this._deflate = zlib.createDeflateRaw({
+        flush: zlib.Z_SYNC_FLUSH,
+        windowBits: typeof maxWindowBits === 'number' ? maxWindowBits : DEFAULT_WINDOW_BITS,
+        memLevel: this._options.memLevel || DEFAULT_MEM_LEVEL
+      });
+    }
+    this._deflate.writeInProgress = true;
+
+    var totalLength = 0;
+    const buffers = [];
+
+    const onData = (data) => {
+      totalLength += data.length;
+      buffers.push(data);
+    };
+
+    const onError = (err) => {
+      cleanup();
+      callback(err);
+    };
+
+    const cleanup = () => {
+      if (!this._deflate) return;
+
+      this._deflate.removeListener('error', onError);
+      this._deflate.removeListener('data', onData);
+      this._deflate.writeInProgress = false;
+
+      if (
+        (fin && this.params[`${endpoint}_no_context_takeover`]) ||
```

Comment:
__[jshint]__
* 'template literal syntax' is only available in ES6 (use 'esversion: 6').

<!-- Comment by [SideCI](https://sideci.com) -->